### PR TITLE
Make `extern "C"` explicit for each function

### PR DESCRIPTION
**Pull request by pinkwah**
This PR removes the `extern "C"` blocks that have been so prevalent in
the codebase. These have been initially implemented incorrectly where
they would also force all external `#include`s to have C linkage as
well. Additionally, not all functions needed to have C linkage -- only
ones used in `ResPrototype`. As such, this PR effectively changes the
meaning of `extern "C"` to "this function is used via CWrap".

### CODE REVIEW

File: libres/lib/include/ert/analysis/update.hpp [hunk 1]
```diff
@@ -1,3 +1,5 @@
+#pragma once
```
Comment (oyvindeide) [inline libres/lib/include/ert/analysis/update.hpp:1]: what is being suppressed?

File: libres/lib/include/ert/enkf/config_parser.hpp [hunk 2]
```diff
@@ -1,22 +0,0 @@
...
-   the Free Software Foundation, either version 3 of the License, or
-   (at your option) any later version.
-
-   ERT is distributed in the hope that it will be useful, but WITHOUT ANY
-   WARRANTY; without even the implied warranty of MERCHANTABILITY or
-   FITNESS FOR A PARTICULAR PURPOSE.
-
-   See the GNU General Public License at <http://www.gnu.org/licenses/gpl.html>
-   for more details.
-*/
```
Comment (oyvindeide) [inline libres/lib/include/ert/enkf/config_parser.hpp:17]: Seems fine to delete this, but perhaps a separate PR? Or commit at least.

File: libres/lib/include/ert/enkf/enkf.hpp [hunk 3]
```diff
@@ -1,28 +0,0 @@
...
-#define SPLASH_LENGTH 8
-const char *splash_text[SPLASH_LENGTH] = {
-    "                  8888888888          888    d8P  8888888888",
-    "                  888                 888   d8P   888       ",
-    "                  888                 888  d8P    888       ",
-    "                  8888888    88888b.  888d88K     8888888   ",
-    "                  888        888 '88b 8888888b    888       ",
-    "                  888        888  888 888  Y88b   888       ",
-    "                  888        888  888 888   Y88b  888       ",
-    "                  8888888888 888  888 888    Y88b 888       "};
```
Comment (oyvindeide) [inline libres/lib/include/ert/enkf/enkf.hpp:28]: How can we delete this? Also, same comment, preferably different PR, or at least different commit.

File: libres/lib/include/ert/analysis/update.hpp (hunk 1 above)
Comment (joakim-hove) [reply libres/lib/include/ert/analysis/update.hpp:1]: Repeated includes into the same translation unit.

**Review by oyvindeide — APPROVED** [review]
A few comments regarding changes that belong in a different PR, other than that this is a welcome change :+1:

File: libres/lib/enkf/gen_obs.cpp [hunk 4]
```diff
@@ -484,9 +484,8 @@ void gen_obs_user_get_with_data_index(const gen_obs_type *gen_obs,
     }
 }
 
-C_USED void gen_obs_update_std_scale(gen_obs_type *gen_obs,
```
Comment (oyvindeide) [inline libres/lib/enkf/gen_obs.cpp:487]: Technically not related to `extern C`

**Comment by oyvindeide** [comment]
Still looks good to me :+1: Did you verify that all the drivers still work? Not sure how they are linked, but think it should be fine.

**Comment by oyvindeide** [comment]
Think the flakyness you are seeing in CI has been fixed on main earlier today

**Comment by pinkwah** [comment]
Jenkins test this please

**Comment by pinkwah** [comment]
Jenkins test required please

**Comment by pinkwah** [comment]
Jenkins test required please

**Pull request merged**